providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,667 @@
+    }
+    return key->settable;
+}
+
+int ossl_hybrid_kmgmt_validate(const void *vctx, int selection, int checktype)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t key_length;
+    unsigned int i;
+    const int pub = (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0;
+    const int priv = (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0;
+
+    if (!ossl_prov_is_running())
+        return 0;
+    if ((key_length = ossl_hybrid_get_key_length_from_ctx(ctx)) == 0)
+        return 0;
+
+    if (pub || priv)
+        for (i = 0; i < ctx->num_algs; i++)
+            if ((priv && !EVP_PKEY_private_check(ctx->ctxs[i]))
+                    || (pub && !EVP_PKEY_public_check(ctx->ctxs[i])))
+                return 0;
+    return 1;
+}
+
+void ossl_hybrid_pkey_ctx_free(HYBRID_PKEY_CTX *ctx)
+{
+    unsigned int i;
+
+    if (ctx != NULL) {
+        for (i = 0; i < ctx->num_algs; i++)
+            EVP_PKEY_CTX_free(ctx->ctxs[i]);
+        OPENSSL_free(ctx->propq);
+        OSSL_PARAM_free(ctx->gettable);
+        OSSL_PARAM_free(ctx->settable);
+        OPENSSL_free(ctx);
+    }
+}
+
+HYBRID_PKEY_CTX *ossl_hybrid_kmgmt_gen_init(void *provctx,
+                                            int selection,
+                                            const OSSL_PARAM params[],
+                                            unsigned int num_algs,
+                                            const char **names,
+                                            uint32_t *key_lengths)
+{
+    OSSL_LIB_CTX *libctx = PROV_LIBCTX_OF(provctx);
+    HYBRID_PKEY_CTX *ctx = NULL;
+    unsigned int i;
+
+    if (!ossl_prov_is_running())
+        return NULL;
+
+    if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) != NULL) {
+        ctx->libctx = libctx;
+        ctx->propq = NULL;
+        ctx->num_algs = num_algs;
+        ctx->names = names;
+        ctx->key_lengths = key_lengths;
+        for (i = 0; i < ctx->num_algs; i++) {
+            ctx->ctxs[i] = EVP_PKEY_CTX_new_from_name(libctx, ctx->names[i],
+                                                      ctx->propq);
+            if (ctx->ctxs[i] == NULL
+                    || !EVP_PKEY_keygen_init(ctx->ctxs[i]))
+                goto err;
+        }
+        if (!ossl_hybrid_kmgmt_gen_set_params(ctx, params))
+            goto err;
+    }
+    return ctx;
+ err:
+    ossl_hybrid_pkey_ctx_free(ctx);
+    return NULL;
+}
+
+int ossl_hybrid_get_ctx_params(HYBRID_PKEY_CTX *ctx, OSSL_PARAM params[])
+{
+    OSSL_PARAM prms[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+    unsigned char *q;
+    OSSL_PARAM *p;
+    uint32_t t;
+    unsigned int i;
+    int type;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    for (p = params; params->key != NULL; params++) {
+        type = ossl_param_find_pidx(p->key);
+        switch (type) {
+        case PIDX_PKEY_PARAM_SECURITY_BITS:
+            t = maximum_ctx_uint32_params(ctx->num_algs, ctx->ctxs,
```

Comment:
Not sure if its the max or the min of the algorithms.

Reply:
It's the max.  If one alg is secure, the hybrid is.